neo/Wallets/Wallet.cs

```diff
@@ -312,11 +313,66 @@ private Transaction MakeTransaction(Snapshot snapshot, byte[] script, Transactio
+                        {
+                            // Try to find the expected results
+
+                            var res = attributes
+                                .Where(u => u.Usage == TransactionAttributeUsage.OracleRequest)
+                                .Cast<OracleExpectedResult>()
+                                .FirstOrDefault();
+
+                            // If there are expected hash, we wil return an empty result
+
+                            oracle = res == null ? null : new OracleExecutionCache((r) =>
+                            {
+                                if (res.ContainsRequest(r.Hash)) return OracleResult.CreateResult(tx.Hash, r.Hash, new byte[0]);
+                                throw new ArgumentException();
+                            });
+                            break;
+                        }
+                    case OracleExecutionType.UnsecureOracle:
+                        {
+                            // All syscalls will be collected, but without the real expected hash
+                            oracle = new OracleExecutionCache((r) => OracleResult.CreateResult(tx.Hash, r.Hash, new byte[0]));
+                            break;
+                        }
+                    case OracleExecutionType.SecureOracle:
+                        {
+                            var service = new OracleService() { TimeOut = TimeSpan.FromSeconds(2) };
+                            oracle = service.CreateExecutionCache(tx.Hash);
+                            break;
+                        }
+                    default: throw new ArgumentException(nameof(oracle));
+                }
+
                 // will try to execute 'transfer' script to check if it works
-                using (ApplicationEngine engine = ApplicationEngine.Run(script, snapshot.Clone(), tx, testMode: true))
+                using (ApplicationEngine engine = ApplicationEngine.Run(script, snapshot.Clone(), tx, testMode: true, oracle: oracle))
                 {
                     if (engine.State.HasFlag(VMState.FAULT))
                         throw new InvalidOperationException($"Failed execution for '{script.ToHexString()}'");
+
+                    if (oracleType != OracleExecutionType.WithoutOracles && oracle.Count > 0)
+                    {
+                        // Attach the OracleExpectedResult attribute
+
+                        var attr = new List<TransactionAttribute>(attributes);
+                        attr.RemoveAll(u => u.Usage == TransactionAttributeUsage.OracleRequest);
+                        attr.Add(new OracleExpectedResult(oracle, oracleType == OracleExecutionType.SecureOracle));
+                        tx.Attributes = attr.ToArray();
+
+                        // Try again, because it possible that the SC use the attributes of the TX
+
+                        return MakeTransaction(snapshot, script, attributes, cosigners, balances_gas, OracleExecutionType.WithoutOracles);
```

Comment:
This is for me the complicated work, the SC can change the required fee according to the downloaded result, and the attributes attached. So maybe we should enable to set the fee as a parameter.